src/python/pants/backend/scala/dependency_inference/ScalaParser.scala

```diff
@@ -209,8 +214,25 @@ class SourceAnalysisTraverser extends Traverser {
       visitMods(mods)
       val name = extractName(nameNode)
       recordScope(name)
-      recordProvidedName(name, sawObject = true)
-      visitTemplate(templ, name)
+
+      // TODO: should object already be recursive?
+      // an object is recursive if extends another type because we cannot figure out the provided types
+      // in the parents, we just mark the object as recursive (which is indicated by non-empty inits)
+      val recursive = !templ.inits.isEmpty
```

Comment:
I'll wait for Stu to comment

Reply:
this piece of code is actually a copy from the `Defn.Object`, so not really new.